File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/JoinerSlot.java
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner;

import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult;
import org.jetbrains.annotations.NotNull;

import java.text.ParseException;
import java.util.regex.Pattern;

public record JoinerSlot(
        MatchTransformer transformer,
        int count,
        boolean sameMatcher,
        String delimiter
) implements MatchTransformer {
    /**
     * GRAMMAR:
     * <pre>
     *     transformer : body? parameters?;
     *     body        : matcher ("|" matcher)*
     *     matcher     : CHAR "-" CHAR | NAME
     *     parameters  : "#" quantifier ("&" delimiter)?
     *     quantifier  : (NUMBER | "*" | "?")
     *     delimiter   : [^]*
     *     NAME        : CHAR+
     *     CHAR        : [^#?|-]+
     *     NUMBER      : [0-9]+
     * </pre>
     */
    private static final Pattern SLOT_EXPR = Pattern.compile(
            "^(?<body>([^#?|-]-[^#?|-]|[^#?|-]+)(\\|([^#?|-]-[^#?|-]|[^#?|-]+))*)?" +
                    "(#(?<count>\\d+|\\*|\\?)(&(?<join>.*))?)?$"
    );

    @Override
    public void transform(@NotNull MatchResult r, @NotNull StringBuilder result) {
        transformer.transform(r, result);
    }

    @Override
    public boolean acceptsWidth(int w) {
        return transformer.acceptsWidth(w);
    }

    @Override
    public int acceptedWidth() {
        return transformer.acceptedWidth();
    }

    public static @NotNull JoinerSlot parse(String content, int start) throws ParseException {
        final var parts = SLOT_EXPR.matcher(content);
        if (!parts.matches()) {
            throw new ParseException("Invalid Slot: " + content, start);
        }
        boolean sameCapture = false;
        final var countStr = parts.group("count");
        final int count;
        if (countStr == null) {
            count = 1;
        } else {
            count = switch (countStr) {
                default -> Integer.parseInt(countStr);
                case "*" -> Integer.MAX_VALUE;
                case "?" -> {
                    sameCapture = true;
                    yield Integer.MAX_VALUE;
                }
            };
        }
        var joinWith = parts.group("join");
        if (joinWith == null) joinWith = "";
        final var body = parts.group("body");
        final MatchTransformer transformer;
        if (body == null) {
            transformer = IdentityTransformer.INSTANCE;
        } else {
            final var values = body.split("\\|");
            final var transformers = new MatchTransformer[values.length];
            for(int i = 0; i < values.length; i++) {
                final var v = values[i];
                final var minus = v.indexOf('-');
                if (minus >= 0) {
                    transformers[i] = new RangeTransformer(v.codePointAt(0), v.codePointAt(minus + 1));
                } else {
                    transformers[i] = new ConstantTransformer(v);
                }
            }
            transformer = new EnumeratedTransformer(transformers);
        }
        return new JoinerSlot(transformer, count, sameCapture, joinWith);
    }
}

File: common/src/main/kotlin/io/github/colintimbarndt/chat_emotes/common/data/unicode/joiner/JoinerSlot.kt
package io.github.colintimbarndt.chat_emotes.common.data.unicode.joiner

import io.github.colintimbarndt.chat_emotes.common.data.unicode.pattern.MatchResult
import java.text.ParseException
import java.util.*
import java.util.regex.Pattern

data class JoinerSlot(
    val transformer: MatchTransformer,
    val count: Int,
    val sameMatcher: Boolean,
    val delimiter: String
) : MatchTransformer {
    override fun transform(r: MatchResult, result: StringBuilder) {
        transformer.transform(r, result)
    }

    override fun acceptsWidth(w: Int): Boolean {
        return transformer.acceptsWidth(w)
    }

    override val acceptedWidth: Int
        get() = transformer.acceptedWidth

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other == null || other.javaClass != this.javaClass) return false
        val that = other as JoinerSlot
        return transformer == that.transformer
                && count == that.count
                && sameMatcher == that.sameMatcher
                && delimiter == that.delimiter
    }

    override fun hashCode(): Int {
        return Objects.hash(transformer, count, sameMatcher, delimiter)
    }

    override fun toString(): String {
        return "JoinerSlot[" +
                "transformer=" + transformer + ", " +
                "count=" + count + ", " +
                "sameMatcher=" + sameMatcher + ", " +
                "delimiter=" + delimiter + ']'
    }

    companion object {
        /**
         * GRAMMAR:
         * <pre>
         * transformer : body? parameters?;
         * body        : matcher ("|" matcher)*
         * matcher     : CHAR "-" CHAR | NAME
         * parameters  : "#" quantifier ("&" delimiter)?
         * quantifier  : (NUMBER | "*" | "?")
         * delimiter   : [^]*
         * NAME        : CHAR+
         * CHAR        : [^#?|-]+
         * NUMBER      : [0-9]+
        </pre> *
         */
        private val SLOT_EXPR = Pattern.compile(
            "^(?<body>([^#?|-]-[^#?|-]|[^#?|-]+)(\\|([^#?|-]-[^#?|-]|[^#?|-]+))*)?" +
                    "(#(?<count>\\d+|\\*|\\?)(&(?<join>.*))?)?$"
        )

        @Throws(ParseException::class)
        fun parse(content: String, start: Int): JoinerSlot {
            val parts = SLOT_EXPR.matcher(content)
            if (!parts.matches()) {
                throw ParseException("Invalid Slot: $content", start)
            }
            var sameCapture = false
            val countStr = parts.group("count")
            val count: Int
            if (countStr == null) {
                count = 1
            } else {
                count = when (countStr) {
                    "*" -> Int.MAX_VALUE
                    "?" -> {
                        sameCapture = true
                        Int.MAX_VALUE
                    }
                    else -> countStr.toInt()
                }
            }
            var joinWith = parts.group("join")
            if (joinWith == null) joinWith = ""
            val body = parts.group("body")
            val transformer: MatchTransformer
            if (body == null) {
                transformer = IdentityTransformer
            } else {
                val values = body.split('|')
                val transformers = arrayOfNulls<MatchTransformer>(values.size)
                for ((i, v) in values.withIndex()) {
                    val minus = v.indexOf('-')
                    if (minus >= 0) {
                        transformers[i] = RangeTransformer(v.codePointAt(0), v.codePointAt(minus + 1))
                    } else {
                        transformers[i] = ConstantTransformer(v)
                    }
                }
                transformer = EnumeratedTransformer(transformers.requireNoNulls())
            }
            return JoinerSlot(transformer, count, sameCapture, joinWith)
        }
    }
}